### ISSUE TITLE
Getting java.io.EOFException when httpClient send request

### ISSUE THREAD

**Issue by jinhong0220** [eclipse/jetty.project#3485]
When httpClient send request to Server, the server receive the request and respond.
But httpClient get result failed. The errors:

```
Caused by: java.util.concurrent.ExecutionException: java.io.EOFException: HttpConnectionOverHTTP@7faba232(l:/127.0.0.1:57718 <-> r:/127.0.0.1:19443,closed=false)[HttpChannelOverHTTP@efe0ede(exchange=HttpExchange@36ffce00 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@6e5da702(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@f46f588{s=START}],recv=HttpReceiverOverHTTP@2d5e6c60(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of 0}]]]
        at org.eclipse.jetty.client.util.FutureResponseListener.getResult(FutureResponseListener.java:118) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.util.FutureResponseListener.get(FutureResponseListener.java:110) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:654) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]

Caused by: java.io.EOFException: HttpConnectionOverHTTP@7faba232(l:/127.0.0.1:57718 <-> r:/127.0.0.1:19443,closed=false)[HttpChannelOverHTTP@efe0ede(exchange=HttpExchange@36ffce00 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@6e5da702(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@f46f588{s=START}],recv=HttpReceiverOverHTTP@2d5e6c60(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of 0}]]]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:292) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1374) ~[org.eclipse.jetty-jetty-http-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:185) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:132) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:70) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:90) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:115) ~[org.eclipse.jetty-jetty-client-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) ~[org.eclipse.jetty-jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) ~[org.eclipse.jetty-jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:202) ~[org.eclipse.jetty-jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:273) ~[org.eclipse.jetty-jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:95) ~[org.eclipse.jetty-jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) ~[org.eclipse.jetty-jetty-io-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) ~[org.eclipse.jetty-jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) ~[org.eclipse.jetty-jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) ~[org.eclipse.jetty-jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) ~[org.eclipse.jetty-jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) ~[org.eclipse.jetty-jetty-util-9.3.11.v20160721.jar:9.3.11.v20160721]
        at java.lang.Thread.run(Thread.java:759) ~[?:1.8.0_151-internal]
```

I'm using jetty version 9.3.11.v20160721

**Comment by sbordet** [comment]
The server did not respond, it closed the connection to the client. Or some intermediary did.

**Comment by radusw** [comment]
I'm experiencing the same issue with jetty. Other http clients do not have this issue, btw. I think it's caused by a race condition.

**Comment by joakime** [comment]
@radusw do you have any additional information?
Perhaps even a pcap of the transaction that caused your case?

**Comment by radusw** [comment]
I have only a stack trace. Why I believe it's a race condition is that it happens about once every 3 runs in my case. I have noticed happening when 2 requests are made in a succession.

```
[error]      java.io.EOFException: HttpConnectionOverHTTP@52204608::SocketChannelEndPoint@1057248d{localhost/127.0.0.1:8080<->/127.0.0.1:53170,ISHUT,fill=-,flush=-,to=36/9223372036854775807}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@52204608(l:/127.0.0.1:53170 <-> r:localhost/127.0.0.1:8080,closed=false)=>HttpChannelOverHTTP@32180f40(exchange=HttpExchange@18808f0a req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@5acee8dd(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@7394dc45{s=START}],recv=HttpReceiverOverHTTP@184bf78(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]] (HttpReceiverOverHTTP.java:338)
[error] org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:338)
[error] org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1552)
[error] org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:209)
[error] org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:147)
[error] org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73)
[error] org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133)
[error] org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155)
[error] org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
[error] org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
[error] org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
[error] org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
[error] org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
[error] org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
[error] org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
[error] org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
[error] org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:781)
[error] org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:917)
```

**Comment by sbordet** [comment]
@radusw you are not using SSL, so it cannot be related to #3891.

The stack trace shows that the server closed the connection.
We need more information as to why the server closed the connection.
